runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/jobsubmission/JobInvocationTest.java

```diff
@@ -130,6 +128,40 @@ public void testNoCancellationWhenDone() throws Exception {
     assertThat(pipelineResult.cancelLatch.getCount(), is(1L));
   }
 
+  @Test(timeout = 10_000)
+  public void testReturnsMetricsFromJobInvocationAfterSuccess() throws Exception {
+    JobApi.MetricResults expectedMonitoringInfos = JobApi.MetricResults.newBuilder().build();
+    TestPipelineResult result =
+        new TestPipelineResult(PipelineResult.State.DONE, expectedMonitoringInfos);
+
+    jobInvocation.start();
+    runner.setResult(result);
+
+    awaitJobState(jobInvocation, JobApi.JobState.Enum.DONE);
+
+    assertThat(
+        jobInvocation.getMetrics(),
+        allOf(is(notNullValue()), is(sameInstance(result.portableMetrics()))));
+  }
+
+  @Test(timeout = 10_000)
+  public void testReturnsMetricsFromJobInvocationAfterCancellation() throws Exception {
+    JobApi.MetricResults expectedMonitoringInfos = JobApi.MetricResults.newBuilder().build();
+    TestPipelineResult result =
+        new TestPipelineResult(PipelineResult.State.RUNNING, expectedMonitoringInfos);
+
+    jobInvocation.start();
+    runner.setResult(result);
+    awaitJobState(jobInvocation, JobApi.JobState.Enum.RUNNING);
+
+    jobInvocation.cancel();
+    awaitJobState(jobInvocation, JobApi.JobState.Enum.CANCELLED);
+
+    assertThat(
```

Comment:
@ibzib, @angoenka  I thought I fixed this test with [the recent commit](https://github.com/apache/beam/pull/9020/commits/ebe630f668d0e83d4445f13402d1a53e35307a6b) but it turns out I didn't. 

However, on the second thought, I think I cannot meet this assertion requirements in this test - if [`onFailure()` callback](https://github.com/apache/beam/blob/de8c987f0cf5749d4913379f0fc3901dc2825192/runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobInvocation.java#L167) gets invoked in `JobInvocation.cancel()`, the metrics cannot be set because we do not have access to pipelineOptions in `onFailure(Throwable throwable)`. I currently suspect that this is the reason why the test is flaky (still investigating). 

I think I have to test this differently. 

Could you share your opinion about this?

Reply:
@ibzib @angoenka 

Update: I know why this test is flaky! 

1. Occasionally, invocationFuture in start() ends up in onFailure method due to a CancellationException and [sets the CANCELLED state](https://github.com/apache/beam/blob/5c6f4282c698c6fca38e576230642eacaf0bea55/runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobInvocation.java#L116).
2. In the test we are waiting for the CANCELLED state but assuming that it will appear only when we actually cancel() the job (and that assumption is incorrect, as point 1 shows). 
3. the metrics are not there yet because neither start's onSuccess() or cancel's onSuccess() were called so the assertion is not satisfied.

I had an idea how to test this:
```
@Test(timeout = 10_000)
  public void testReturnsMetricsFromJobInvocationAfterCancellation() throws Exception {
    JobApi.MetricResults expectedMonitoringInfos = JobApi.MetricResults.newBuilder().build();
    TestPipelineResult result =
        new TestPipelineResult(PipelineResult.State.RUNNING, expectedMonitoringInfos);

    runner.setResult(result);
    invocationFuture = executorService.submit(() -> { runner.run(pipeline, jobInfo);});
    jobInvocation.cancel();
    awaitJobState(jobInvocation, JobApi.JobState.Enum.CANCELLED);

    assertThat(
        jobInvocation.getMetrics(),
        allOf(is(notNullValue()), is(sameInstance(result.portableMetrics()))));
  }
```

BUT it's currently not doable without using any sort of reflection (which I don't want to do because this would be nasty) - I don't have access to `invocationFuture` in `JobInvocationTest`. 

Let me know if you have any comments or ideas on how to fix this differently. I currently intend to **delete this flaky test**. Let me know if you agree to that.

Reply:
example of a failure: https://builds.apache.org/job/beam_PreCommit_Java_Phrase/1181/

Reply:
I guess the important question here is, can the user expect metrics to be returned if the pipeline is cancelled? If yes, you should find a way to keep the test (maybe by marking `invocationFuture` as `@VisibleForTesting protected`). If no, the test flakiness is indicative of an unreliable feature, which we should either fix, eliminate, or otherwise indicate to the user somehow.

Reply:
Yes - that is a good question indeed. I raised that on the devlist - let's see what the opinions are.

Reply:
Returning metrics when cancelled is quite normal, at least in the dataflow case. A user may cancel a pipeline due to some stuckness or slowness issue they need to debug. Then they browse the DF UI to debug, the UI contains metrics and other debugging info.